### ISSUE TITLE
Script: Engagement Scaffolding

### ISSUE THREAD

**Issue by fullstackpotato** [stevemcilwain/quiver#12]
As suggested in the previous comment, please find a sanitized version of my engagement scaffolding script and a cleanup one.

Suggested Filename: `qq-engagement.zsh`
Aliases: `qq-engagement-start/qq-engagement-end`
Required tools: `tree`,`7zip`

## Engagement Scaffolding

Pretty simple start and end script that I use for creating consistent folder layouts for tests. Be aware that I do have some extra pieces within these functions that I have removed to put online.

They're very basic and not the most elegant solutions but appear to work reliably.

**Used variables**

I exported 3 variables as I use them elsewhere with some internal tools/scripts. These variables are:

```bash
export __ENGAGEMENTSROOT="/engagement/"
export __CONSULTANTNAME="stevemcilwain"
export __CONSULTANTEMAIL="stevemcilwain@emailprovider.com"
```

## Start Engagement

This will:

1. Request an engagement name and ID
2. make folder structure (using details from above)
3. fold in your qq-log-set to get my notes file set right.

```bash
qq-engagement-start() {
    local id && read "id?Engagement ID: "
    local n && read "n?Engagement Name: "

    local choice && read "choice?Continue (y/n)?: "
    case "$choice" in 
        y|Y ) 
            local d=`date '+%d-%m-%Y'`
            #__ED = Engagement Directory: id-EngagementName-ConsultantName
            export __ED="${id}-${n}-${__CONSULTANTNAME// /}"
	    # create folders for engagements
            mkdir -p ${__ENGAGEMENTSROOT}${__ED}/{burp/{log,intruder,http-requests},client-supplied-info/emails,files/{downloads,uploads},notes/screenshots,scans/{raw,pretty},ssl,tool-output}
            
            # Have left this in to show what the extra steps might entail, may want removing if folded into quiver
            
            # copy my engagement template files
            cp /templates/engagement/* ${__ENGAGEMENTSROOT}${__ED}/notes/
	    # Have removed some sed's that insert specific text into my note template files
			
            ;;
        n|N ) 
            echo "no"
            ;;
        * ) 
            echo "invalid"
            ;;
    esac

    # wanted this to be an optional step, sometimes I'll create folders in advance due to calls with clients ahead of the test or prep work
    local setlog && read "setlog?Set log file to this engagement (y/n)?: "
    case "$setlog" in 
        y|Y ) 
            print -z "qq-log-set ${__ENGAGEMENTSROOT}${__ED}/notes/engagement-notes.md"
            ;;
        n|N ) 
            echo "no"
            ;;
        * ) 
            echo "invalid"
            ;;
    esac
}
```

## End engagement

This will:

1. delete empty/unused folders
2. create a tree of the directory as a HTML
3. zip the engagement folder up
4. give the option to delete the folder

```bash
qq-engagement-end() {
    cd "$__ENGAGEMENTSROOT"
    printf "? Please select folder:\n"
    select d in */; do test -n "$d" && break; echo ">>> Invalid Selection"; done
    cd "$d" # jump to selected folder

    local sd=$(pwd) # $d wasn't working, I'm not a smart man. aka workaround!

    # Task 1: delete all empty folders
    local delfol && read "delfol?Delete empty folders? (Y/n)?: "
    if [[ "$delfol" =~ ^[Yy]$ ]]
    then
        find . -type d -empty -delete # delete empty folders
    fi

    # Task 2: create tree
    tree -C -F -H ./ > "${sd}/tree.html" || echo -e "! failed creating tree..."

    # Task 3: zip up engagement folder
    local zf=$(basename ${sd})
    7z a -t7z -m0=lzma2 -mx=9 -mfb=64 -md=1024m -ms=on ${__ENGAGEMENTSROOT}/${zf}.7z ${sd} || echo -e "! failed zipping ${sd} ${zf}" # or condition to report fails

    # Task 4: Delete engagement folder
    local delenfol && read "delenfol?Delete engagement folder? (Y/n)?: "
    if [[ "$delenfol" =~ ^[Yy]$ ]]
    then
        rm -rf $sd # delete unwanted engagement folder
    fi
}
```

**Comment by stevemcilwain** [comment]
I have a lot of "qq-enum" namespaces.  I'd prefer to this as qq-project instead?  Also makes it a little more generic as I'd like this framework to fit for both pen test and bug bounty.

**Comment by stevemcilwain** [comment]
@zneedotdev what's the purpose of the "ssl" directory? Discovered certs?

**Comment by fullstackpotato** [comment]
@stevemcilwain I quickly moved it to engagement as my previous bash alias was `engagement-start`. I have found that engagement and enum being a bit of a pain for tab completion. Project would be so much easier!

Re ssl: I have the ssl directory to save certificates, output all formats for tools such as testssl and an internal ssl tool that my company has. I have found that on large infrastructure tests that these files can take over the tool-output folder so it made sense to me to make the ssl folder to keep it separate. I can then parse files in these folders for details/supporting materials with some bash one-liners. This might be different for everyone else and I accept that I could well be a niche case.

**Comment by stevemcilwain** [comment]
So @zneedotdev  I went a little crazy with this stuff... what started as a simple scaffolding thing turned into me adding a qq-vars.zsh namespace with global variables for things like output directory, domain, url, rhost, lhost, network, wordlist and more (which you can load and save).  Then I put scaffolding dirs in qq-vars-set-output and weaved in qq-log.  I wired up all the functions with outputs to $__OUTPUT.  I know it's probably not quite what you want, so I'll still add your specifics to engagement and try to make that work in parallel.

If you get a chance try the dev branch:  git clone -b DEV-0.12 git@github.com:stevemcilwain/quiver.git

I am still testing and finishing this massive branch, but so far I have found it to speed up my workflow.  Let me know what you think.

**Comment by stevemcilwain** [comment]
OK I have a fairly simple, usable version of qq-project.zsh:

qq-project-zd-start
qq-project-zd-end

Named after you so we can have alternate versions of scaffolding ; )

**Comment by fullstackpotato** [comment]
@stevemcilwain Absolutely amazing work. Things have been busy here in the UK with the whole Covid-19 stuff and adjusting to wfh. Hope you and your family are staying safe!
I had a quick look at the scripts, you really weren't joking about going deep with the changes. They look really good. I'm always flexible and interested in trying new ways, I'll give it a try over this weekend/into next week.

And amazing, thank you for the custom ones :)

**Comment by stevemcilwain** [comment]
Merged
